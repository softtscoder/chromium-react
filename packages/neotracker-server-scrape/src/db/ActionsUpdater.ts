import { addressToScriptHash, RawAction } from '@neo-one/client-full';
import { createChild, serverLogger } from '@neotracker/logger';
import { Action as ActionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { strip0x } from '../utils';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface ActionsSaveSingle {
  readonly action: RawAction;
  readonly transactionID: string;
  readonly transactionHash: string;
}
export interface ActionsSave {
  readonly actions: ReadonlyArray<ActionsSaveSingle>;
}
export interface ActionsRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class ActionsUpdater extends SameContextDBUpdater<ActionsSave, ActionsRevert> {
  public async save(context: Context, { actions }: ActionsSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_actions' });
    await Promise.all(
      _.chunk(actions, context.chunkSize).map(async (chunk) =>
        ActionModel.insertAll(
          context.db,
          context.makeQueryContext(),
          chunk.map(({ action, transactionID, transactionHash }) => ({
            id: action.globalIndex.toString(),
            type: action.type,
            block_id: action.blockIndex,
            transaction_id: transactionID,
            transaction_hash: transactionHash,
            transaction_index: action.transactionIndex,
            index: action.index,
            script_hash: strip0x(addressToScriptHash(action.address)),
            // tslint:disable no-any no-null-keyword
            message: (action as any).message === undefined ? null : encodeURIComponent((action as any).message),
            args_raw: (action as any).args === undefined ? null : JSON.stringify((action as any).args),
            // tslint:disable-enable no-any no-null-keyword
          })),
        ),
      ),
    );
  }

  public async revert(context: Context, { transactionIDs }: ActionsRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_actions' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map(async (chunk) =>
        ActionModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('transaction_id', chunk)
          .delete(),
      ),
    );
  }
}
