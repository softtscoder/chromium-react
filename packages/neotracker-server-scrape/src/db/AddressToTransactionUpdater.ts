import { createChild, serverLogger } from '@neotracker/logger';
import { AddressToTransaction as AddressToTransactionModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressToTransactionSaveSingle {
  readonly addressIDs: ReadonlyArray<string>;
  readonly transactionID: string;
}
export interface AddressToTransactionSave {
  readonly transactions: ReadonlyArray<AddressToTransactionSaveSingle>;
}
export interface AddressToTransactionRevert {
  readonly transactionIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressToTransactionUpdater extends SameContextDBUpdater<
  AddressToTransactionSave,
  AddressToTransactionRevert
> {
  public async save(context: Context, { transactions }: AddressToTransactionSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_address_to_transaction' });
    const data = _.flatMap(transactions, ({ addressIDs, transactionID }) =>
      [...new Set(addressIDs)].map((addressID) => ({
        id1: addressID,
        id2: transactionID,
      })),
    );
    await Promise.all(
      _.chunk(data, context.chunkSize).map(async (chunk) => {
        await AddressToTransactionModel.insertAll(context.db, context.makeQueryContext(), chunk);
      }),
    );
  }

  public async revert(context: Context, { transactionIDs }: AddressToTransactionRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_address_to_transaction' });
    await Promise.all(
      _.chunk(transactionIDs, context.chunkSize).map((chunk) =>
        AddressToTransactionModel.query(context.db)
          .context(context.makeQueryContext())
          .delete()
          .whereIn('id2', chunk),
      ),
    );
  }
}
