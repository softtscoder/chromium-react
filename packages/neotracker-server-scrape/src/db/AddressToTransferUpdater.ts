import { createChild, serverLogger } from '@neotracker/logger';
import { AddressToTransfer as AddressToTransferModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressToTransferSaveSingle {
  readonly addressIDs: ReadonlyArray<string>;
  readonly transferID: string;
}
export interface AddressToTransferSave {
  readonly transfers: ReadonlyArray<AddressToTransferSaveSingle>;
}
export interface AddressToTransferRevert {
  readonly transferIDs: ReadonlyArray<string>;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressToTransferUpdater extends SameContextDBUpdater<AddressToTransferSave, AddressToTransferRevert> {
  public async save(context: Context, { transfers }: AddressToTransferSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_address_to_transfer' });
    const data = _.flatMap(transfers, ({ addressIDs, transferID }) =>
      [...new Set(addressIDs)].map((addressID) => ({
        id1: addressID,
        id2: transferID,
      })),
    );
    await Promise.all(
      _.chunk(data, context.chunkSize).map(async (chunk) => {
        await AddressToTransferModel.insertAll(context.db, context.makeQueryContext(), chunk);
      }),
    );
  }

  public async revert(context: Context, { transferIDs }: AddressToTransferRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_address_to_transfer' });
    await Promise.all(
      _.chunk(transferIDs, context.chunkSize).map(async (chunk) => {
        await AddressToTransferModel.query(context.db)
          .context(context.makeQueryContext())
          .delete()
          .whereIn('id2', chunk);
      }),
    );
  }
}
