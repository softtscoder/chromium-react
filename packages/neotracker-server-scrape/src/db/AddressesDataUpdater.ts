import { createChild, serverLogger } from '@neotracker/logger';
import { Address as AddressModel, isPostgres } from '@neotracker/server-db';
import { raw } from 'objection';
import { Addresses, Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressesDataSave {
  readonly addresses: Addresses;
  readonly blockIndex: number;
  readonly blockTime: number;
}
export interface AddressesDataRevert {
  readonly addresses: Addresses;
  readonly transactionIDs: ReadonlyArray<string>;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressesDataUpdater extends SameContextDBUpdater<AddressesDataSave, AddressesDataRevert> {
  public async save(context: Context, { addresses, blockIndex, blockTime }: AddressesDataSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(
        async ([address, { transactionCount, transferCount, transactionID, transactionHash }]) => {
          await AddressModel.query(context.db)
            .context(context.makeQueryContext())
            .where('id', address)
            .where('aggregate_block_id', '<', blockIndex)
            .patch({
              // tslint:disable no-any
              transaction_count: raw(`transaction_count + ${transactionCount}`) as any,
              transfer_count: raw(`transfer_count + ${transferCount}`) as any,
              // tslint:enable no-any
              last_transaction_id: transactionID,
              last_transaction_hash: transactionHash,
              last_transaction_time: blockTime,
              aggregate_block_id: blockIndex,
            });
        },
      ),
    );
  }

  public async revert(context: Context, { addresses, transactionIDs, blockIndex }: AddressesDataRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_addresses_data' });
    await Promise.all(
      Object.entries(addresses).map(async ([address, { transactionCount, transferCount }]) => {
        await AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', address)
          .where('aggregate_block_id', '>=', blockIndex)
          .patch({
            // tslint:disable no-any
            transaction_count: raw(`transaction_count - ${transactionCount}`) as any,
            transfer_count: raw(`transfer_count - ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex - 1,
          });
      }),
    );

    const addressIDsSet = Object.keys(addresses);
    if (addressIDsSet.length > 0) {
      if (isPostgres(context.db)) {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.id AS address_id, b.transaction_id, b.hash, b.block_time
                FROM address a
                LEFT OUTER JOIN (
                  SELECT address_id, transaction_id, hash, block_time
                  FROM (
                    SELECT id1 AS address_id, MAX(id2) AS transaction_id
                    FROM address_to_transaction
                    WHERE
                      id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                      id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                    GROUP BY address_to_transaction.id1
                  ) a
                  JOIN transaction b ON
                    a.transaction_id = b.id
                ) b ON
                  a.id = b.address_id
                WHERE
                  a.id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
              )
              UPDATE address SET
                last_transaction_id=cte.transaction_id,
                last_transaction_hash=cte.hash,
                last_transaction_time=cte.block_time
              FROM cte
              WHERE
                id = cte.address_id
              `,
          )
          .queryContext(context.makeQueryContext());
      } else {
        await context.db
          .raw(
            `
              WITH cte AS (
                SELECT a.address_id, transaction_id, hash, block_time
                FROM (
                  SELECT id1 AS address_id, MAX(id2) AS transaction_id
                  FROM address_to_transaction
                  WHERE
                    id1 IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')}) AND
                    id2 NOT IN (${transactionIDs.map((id) => `'${id}'`).join(', ')})
                  GROUP BY id1
                ) a
                JOIN 'transaction' b ON
                  a.transaction_id = b.id
              )
              UPDATE address SET
                last_transaction_id=(select cte.transaction_id from cte where address_id = address.id),
                last_transaction_hash=(select cte.hash from cte where address_id = address.id),
                last_transaction_time=(select cte.block_time from cte where address_id = address.id)
              WHERE
                id IN (${addressIDsSet.map((id) => `'${id}'`).join(', ')})
                `,
          )
          .queryContext(context.makeQueryContext());
      }
    }
  }
}
