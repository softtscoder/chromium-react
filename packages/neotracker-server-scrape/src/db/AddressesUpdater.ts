import { createChild, serverLogger } from '@neotracker/logger';
import { Address as AddressModel } from '@neotracker/server-db';
import _ from 'lodash';
import { Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AddressesSave {
  readonly addresses: ReadonlyArray<Partial<AddressModel> & { readonly id: string }>;
}
export interface AddressesRevert {
  readonly addresses: ReadonlyArray<{ readonly id: string; readonly transactionID?: string }>;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AddressesUpdater extends SameContextDBUpdater<AddressesSave, AddressesRevert> {
  public async save(context: Context, { addresses }: AddressesSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_addresses' });
    const existingAddressIDs = await Promise.all(
      _.chunk(addresses, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map(({ id }) => id)),
      ),
    ).then((result) => _.flatMap(result).map((address) => address.id));

    const existingAddressIDsSet = new Set(existingAddressIDs);
    const toCreate = addresses.filter((address) => !existingAddressIDsSet.has(address.id));

    await Promise.all(
      _.chunk(toCreate, context.chunkSize).map(async (chunk) =>
        AddressModel.insertAll(context.db, context.makeQueryContext(), chunk),
      ),
    );
  }

  public async revert(context: Context, { addresses, blockIndex }: AddressesRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_addresses' });
    const addressModels = await Promise.all(
      _.chunk(addresses, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map(({ id }) => id)),
      ),
    ).then((result) => _.flatMap(result));

    const addressToTransactionID = _.fromPairs(addresses.map(({ id, transactionID }) => [id, transactionID]));
    const toDelete = addressModels.filter(
      (address) =>
        (address.transaction_id == undefined && address.block_id === blockIndex) ||
        (address.transaction_id != undefined && address.transaction_id === addressToTransactionID[address.id]),
    );

    await Promise.all(
      _.chunk(toDelete, context.chunkSize).map((chunk) =>
        AddressModel.query(context.db)
          .context(context.makeQueryContext())
          .whereIn('id', chunk.map((address) => address.id))
          .delete(),
      ),
    );
  }
}
