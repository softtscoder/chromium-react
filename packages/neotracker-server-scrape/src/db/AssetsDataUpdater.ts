import { createChild, serverLogger } from '@neotracker/logger';
import { Asset as AssetModel } from '@neotracker/server-db';
import { raw } from 'objection';
import { Assets, Context } from '../types';
import { SameContextDBUpdater } from './SameContextDBUpdater';

export interface AssetsDataSave {
  readonly assets: Assets;
  readonly blockIndex: number;
}
export interface AssetsDataRevert {
  readonly assets: Assets;
  readonly blockIndex: number;
}

const serverScrapeLogger = createChild(serverLogger, { component: 'scrape' });

export class AssetsDataUpdater extends SameContextDBUpdater<AssetsDataSave, AssetsDataRevert> {
  public async save(context: Context, { assets, blockIndex }: AssetsDataSave): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_save_assets' });
    await Promise.all(
      Object.entries(assets).map(async ([asset, { issued, transactionCount, addressCount, transferCount }]) => {
        await AssetModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', asset)
          .where('aggregate_block_id', '<', blockIndex)
          .patch({
            // tslint:disable no-any
            issued: raw(`issued + (${issued.toString()})`) as any,
            transaction_count: raw(`transaction_count + ${transactionCount}`) as any,
            address_count: raw(`address_count + ${addressCount}`) as any,
            transfer_count: raw(`transfer_count + ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex,
          });
      }),
    );
  }

  public async revert(context: Context, { assets, blockIndex }: AssetsDataRevert): Promise<void> {
    serverScrapeLogger.info({ title: 'neotracker_scrape_revert_assets' });
    await Promise.all(
      Object.entries(assets).map(async ([asset, { issued, transactionCount, addressCount, transferCount }]) => {
        await AssetModel.query(context.db)
          .context(context.makeQueryContext())
          .where('id', asset)
          .where('aggregate_block_id', '>=', blockIndex)
          .patch({
            // tslint:disable no-any
            issued: raw(`issued - (${issued.toString()})`) as any,
            transaction_count: raw(`transaction_count - ${transactionCount}`) as any,
            address_count: raw(`address_count - ${addressCount}`) as any,
            transfer_count: raw(`transfer_count - ${transferCount}`) as any,
            // tslint:enable no-any
            aggregate_block_id: blockIndex - 1,
          });
      }),
    );
  }
}
