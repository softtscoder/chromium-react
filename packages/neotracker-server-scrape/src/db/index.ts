export * from './BlockUpdater';
export * from './utils';
