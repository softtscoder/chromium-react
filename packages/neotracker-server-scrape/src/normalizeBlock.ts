import {
  Asset,
  AssetType,
  Attribute,
  Block,
  ConfirmedTransaction,
  Contract,
  ContractParameter,
  Input,
  Output,
  RawAction,
  RawInvocationData,
  RawInvocationResult,
} from '@neo-one/client-full';
import { utils } from '@neotracker/shared-utils';

export const normalizeHash = (hash: string): string => {
  if (hash.startsWith('0x')) {
    return hash.substring(2);
  }

  return hash;
};

const normalizeInput = (input: Input): Input => ({
  hash: normalizeHash(input.hash),
  index: input.index,
});

const normalizeOutput = (output: Output): Output => ({
  asset: normalizeHash(output.asset),
  value: output.value,
  address: output.address,
});

const normalizeAttribute = (attribute: Attribute): Attribute => ({
  // tslint:disable-next-line no-any
  usage: attribute.usage as any,
  data: normalizeHash(attribute.data),
});

const normalizeContract = (contract: Contract): Contract => ({
  version: contract.version,
  address: contract.address,
  script: contract.script,
  parameters: contract.parameters,
  returnType: contract.returnType,
  name: contract.name,
  codeVersion: contract.codeVersion,
  author: contract.author,
  email: contract.email,
  description: contract.description,
  storage: contract.storage,
  dynamicInvoke: contract.dynamicInvoke,
  payable: contract.payable,
});

const normalizeContractParameter = (contractParameter: ContractParameter): ContractParameter => {
  switch (contractParameter.type) {
    case 'Signature':
      return {
        type: 'Signature',
        value: contractParameter.value,
      };

    case 'Boolean':
      return {
        type: 'Boolean',
        value: contractParameter.value,
      };

    case 'Integer':
      return {
        type: 'Integer',
        value: contractParameter.value,
      };

    case 'Address':
      return {
        type: 'Address',
        value: contractParameter.value,
      };

    case 'Hash256':
      return {
        type: 'Hash256',
        value: normalizeHash(contractParameter.value),
      };

    case 'Buffer':
      return {
        type: 'Buffer',
        value: contractParameter.value,
      };

    case 'PublicKey':
      return {
        type: 'PublicKey',
        value: contractParameter.value,
      };

    case 'String':
      return {
        type: 'String',
        value: contractParameter.value,
      };

    case 'Array':
      return {
        type: 'Array',
        value: contractParameter.value,
      };

    case 'InteropInterface':
      return {
        type: 'InteropInterface',
      };

    case 'Void':
      return {
        type: 'Void',
      };

    case 'Map':
      return {
        type: 'Map',
        value: contractParameter.value,
      };

    default:
      utils.assertNever(contractParameter);
      throw new Error('For TS');
  }
};

export const normalizeAction = (action: RawAction): RawAction => {
  switch (action.type) {
    case 'Log':
      return {
        type: 'Log',
        version: action.version,
        blockIndex: action.blockIndex,
        blockHash: normalizeHash(action.blockHash),
        transactionIndex: action.transactionIndex,
        transactionHash: normalizeHash(action.transactionHash),
        index: action.index,
        globalIndex: action.globalIndex,
        address: action.address,
        message: action.message,
      };

    case 'Notification':
      return {
        type: 'Notification',
        version: action.version,
        blockIndex: action.blockIndex,
        blockHash: normalizeHash(action.blockHash),
        transactionIndex: action.transactionIndex,
        transactionHash: normalizeHash(action.transactionHash),
        index: action.index,
        globalIndex: action.globalIndex,
        address: action.address,
        args: action.args.map(normalizeContractParameter),
      };

    default:
      utils.assertNever(action);
      throw new Error('Unknown action type');
  }
};

const normalizeInvocationResult = (result: RawInvocationResult): RawInvocationResult => {
  switch (result.state) {
    case 'HALT':
      return {
        state: 'HALT',
        gasCost: result.gasCost,
        gasConsumed: result.gasConsumed,
        stack: result.stack.map(normalizeContractParameter),
      };
    case 'FAULT':
      return {
        state: 'FAULT',
        gasCost: result.gasCost,
        gasConsumed: result.gasConsumed,
        stack: result.stack.map(normalizeContractParameter),
        message: result.message,
      };
    default:
      utils.assertNever(result);
      throw new Error('Unknown InvocationResult type');
  }
};

const normalizeAssetType = (asset: AssetType) => {
  switch (asset) {
    case 'Credit':
      return 'CreditFlag';
    case 'Duty':
      return 'DutyFlag';
    case 'Governing':
      return 'GoverningToken';
    case 'Utility':
      return 'UtilityToken';
    case 'Currency':
      return 'Currency';
    case 'Share':
      return 'Share';
    case 'Invoice':
      return 'Invoice';
    case 'Token':
      return 'Token';
    default:
      utils.assertNever(asset);
      throw new Error('Unknown AssetType');
  }
};

const normalizeAsset = (asset: Asset): Asset => ({
  hash: normalizeHash(asset.hash),
  // tslint:disable-next-line no-any
  type: normalizeAssetType(asset.type) as any,
  name: asset.name,
  amount: asset.amount,
  available: asset.available,
  precision: asset.precision,
  owner: asset.owner,
  admin: asset.admin,
  issuer: asset.issuer,
  expiration: asset.expiration,
  frozen: asset.frozen,
});

const normalizeInvocationData = (data: RawInvocationData): RawInvocationData => ({
  result: normalizeInvocationResult(data.result),
  asset: data.asset === undefined ? data.asset : normalizeAsset(data.asset),
  contracts: data.contracts.map(normalizeContract),
  deletedContractAddresses: data.deletedContractAddresses,
  migratedContractAddresses: data.migratedContractAddresses,
  actions: data.actions.map(normalizeAction),
  storageChanges: data.storageChanges,
});

const normalizeTransaction = (transaction: ConfirmedTransaction): ConfirmedTransaction => {
  const transactionBase = {
    hash: normalizeHash(transaction.hash),
    size: transaction.size,
    version: transaction.version,
    attributes: transaction.attributes.map(normalizeAttribute),
    inputs: transaction.inputs.map(normalizeInput),
    outputs: transaction.outputs.map(normalizeOutput),
    scripts: transaction.scripts,
    systemFee: transaction.systemFee,
    networkFee: transaction.networkFee,
    receipt: {
      ...transaction.receipt,
      blockHash: normalizeHash(transaction.receipt.blockHash),
    },
  };

  switch (transaction.type) {
    case 'MinerTransaction':
      return {
        ...transactionBase,
        type: 'MinerTransaction',
        nonce: transaction.nonce,
      };

    case 'IssueTransaction':
      return {
        ...transactionBase,
        type: 'IssueTransaction',
      };

    case 'ClaimTransaction':
      return {
        ...transactionBase,
        type: 'ClaimTransaction',
        claims: transaction.claims.map(normalizeInput),
      };

    case 'EnrollmentTransaction':
      return {
        ...transactionBase,
        type: 'EnrollmentTransaction',
        publicKey: transaction.publicKey,
      };

    case 'RegisterTransaction':
      return {
        ...transactionBase,
        type: 'RegisterTransaction',
        asset: {
          ...transaction.asset,
          // tslint:disable-next-line no-any
          type: normalizeAssetType(transaction.asset.type) as any,
        },
      };

    case 'ContractTransaction':
      return {
        ...transactionBase,
        type: 'ContractTransaction',
      };

    case 'PublishTransaction':
      return {
        ...transactionBase,
        type: 'PublishTransaction',
        contract: normalizeContract(transaction.contract),
      };

    case 'InvocationTransaction':
      return {
        ...transactionBase,
        type: 'InvocationTransaction',
        script: transaction.script,
        gas: transaction.gas,
        invocationData: normalizeInvocationData(transaction.invocationData),
      };

    case 'StateTransaction':
      return {
        ...transactionBase,
        type: 'StateTransaction',
      };

    default:
      utils.assertNever(transaction);
      throw new Error('Unknown transaction type');
  }
};

export const normalizeBlock = (block: Block): Block => ({
  hash: normalizeHash(block.hash),
  size: block.size,
  version: block.version,
  previousBlockHash: normalizeHash(block.previousBlockHash),
  merkleRoot: normalizeHash(block.merkleRoot),
  time: block.time,
  index: block.index,
  nonce: block.nonce,
  nextConsensus: block.nextConsensus,
  script: block.script,
  transactions: block.transactions.map(normalizeTransaction),
});
