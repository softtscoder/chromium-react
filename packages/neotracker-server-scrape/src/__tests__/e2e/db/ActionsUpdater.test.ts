import { RawAction } from '@neo-one/client-full';
import { Database, getDBData, startDB } from '@neotracker/server-test';
import BigNumber from 'bignumber.js';
import Knex from 'knex';
import { ActionsUpdater } from '../../../db/ActionsUpdater';
import { normalizeAction } from '../../../normalizeBlock';
import { data, makeContext } from '../../data';

const createAction = (actionIn: RawAction, transactionID: string) => {
  const action = normalizeAction(actionIn);

  return {
    action,
    transactionID,
    transactionHash: action.transactionHash,
  };
};

describe('ActionsUpdater', () => {
  let database: Database;
  let db: Knex;

  beforeAll(async () => {
    const databaseConfig = await startDB();
    database = databaseConfig.database;
    db = database.knex;
  });

  beforeEach(async () => {
    await database.reset();
  });

  test('inserts actions', async () => {
    const context = makeContext({ db });
    const updater = new ActionsUpdater();

    await updater.save(context, {
      actions: [
        createAction(data.createRawLog({ index: 0, globalIndex: new BigNumber(0) }), '0'),
        createAction(data.createRawNotification({ index: 1, globalIndex: new BigNumber(1) }), '0'),
      ],
    });

    const dbData = await getDBData(db);
    expect(dbData).toMatchSnapshot();
  });

  test('handles duplicate inserts', async () => {
    const context = makeContext({ db });
    const updater = new ActionsUpdater();
    const actions = [
      createAction(data.createRawLog({ index: 0, globalIndex: new BigNumber(0) }), '0'),
      createAction(data.createRawNotification({ index: 1, globalIndex: new BigNumber(1) }), '0'),
    ];

    await updater.save(context, { actions });

    const dbData = await getDBData(db);

    await updater.save(context, { actions });

    const finalDBData = await getDBData(db);
    expect(dbData).toEqual(finalDBData);
  });

  test('reverts actions', async () => {
    const context = makeContext({ db });
    const updater = new ActionsUpdater();

    await updater.save(context, {
      actions: [
        createAction(data.createRawLog({ index: 0, globalIndex: new BigNumber(0) }), '0'),
        createAction(data.createRawNotification({ index: 1, globalIndex: new BigNumber(1) }), '0'),
      ],
    });

    data.nextBlock();
    const dbData = await getDBData(db);

    await updater.save(context, {
      actions: [
        createAction(data.createRawLog({ index: 0, globalIndex: new BigNumber(2) }), '1'),
        createAction(data.createRawNotification({ index: 1, globalIndex: new BigNumber(3) }), '1'),
      ],
    });

    const nextDBData = await getDBData(db);
    expect(nextDBData).toMatchSnapshot();

    await updater.revert(context, { transactionIDs: ['1'] });

    const finalDBData = await getDBData(db);
    expect(dbData).toEqual(finalDBData);
  });
});
