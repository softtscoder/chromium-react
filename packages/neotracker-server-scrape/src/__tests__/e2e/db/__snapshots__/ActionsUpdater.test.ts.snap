// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionsUpdater inserts actions 1`] = `
Object {
  "action": Array [
    Object {
      "args_raw": null,
      "block_id": 1,
      "id": "0",
      "index": 0,
      "message": "foobar",
      "script_hash": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "0",
      "transaction_index": 2,
      "type": "Log",
    },
    Object {
      "args_raw": "[{\\"type\\":\\"String\\",\\"value\\":\\"foobar\\"}]",
      "block_id": 1,
      "id": "1",
      "index": 1,
      "message": null,
      "script_hash": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "0",
      "transaction_index": 2,
      "type": "Notification",
    },
  ],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [],
  "transfer": Array [],
}
`;

exports[`ActionsUpdater reverts actions 1`] = `
Object {
  "action": Array [
    Object {
      "args_raw": null,
      "block_id": 1,
      "id": "0",
      "index": 0,
      "message": "foobar",
      "script_hash": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "0",
      "transaction_index": 2,
      "type": "Log",
    },
    Object {
      "args_raw": "[{\\"type\\":\\"String\\",\\"value\\":\\"foobar\\"}]",
      "block_id": 1,
      "id": "1",
      "index": 1,
      "message": null,
      "script_hash": "ab38352559b8b203bde5fddfa0b07d8b25111111",
      "transaction_hash": "c6c9cd5cacfaa18921cb0869945fbdeda2d0308f3a7458f7a1c9528c06150cf5",
      "transaction_id": "0",
      "transaction_index": 2,
      "type": "Notification",
    },
    Object {
      "args_raw": null,
      "block_id": 2,
      "id": "2",
      "index": 0,
      "message": "foobar",
      "script_hash": "ab38352559b8b203bde5fddfa0b07d8b25333333",
      "transaction_hash": "83ec19a0492cd9146e5f9c6ef278fb66a90a8507f812b2f3901523a300b76b60",
      "transaction_id": "1",
      "transaction_index": 1,
      "type": "Log",
    },
    Object {
      "args_raw": "[{\\"type\\":\\"String\\",\\"value\\":\\"foobar\\"}]",
      "block_id": 2,
      "id": "3",
      "index": 1,
      "message": null,
      "script_hash": "ab38352559b8b203bde5fddfa0b07d8b25333333",
      "transaction_hash": "83ec19a0492cd9146e5f9c6ef278fb66a90a8507f812b2f3901523a300b76b60",
      "transaction_id": "1",
      "transaction_index": 1,
      "type": "Notification",
    },
  ],
  "address": Array [],
  "addressToTransaction": Array [],
  "addressToTransfer": Array [],
  "asset": Array [],
  "assetToTransaction": Array [],
  "block": Array [],
  "coin": Array [],
  "contract": Array [],
  "dataPoint": Array [],
  "migration": Array [],
  "processedIndex": Array [],
  "transaction": Array [],
  "transactionInputOutput": Array [],
  "transfer": Array [],
}
`;
